pilot/pkg/proxy/envoy/v2/internalgen.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"istio.io/istio/pilot/pkg/networking/util"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/golang/protobuf/ptypes/any"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+const (
+	TypeURLConnections = "istio.io/connections"
+	TypeURLDisconnect  = "istio.io/disconnect"
+
+	// TODO: TypeURLReady - readiness events for endpoints, agent can propagate
+
+	// TypeURLNACK will receive messages of type DiscoveryRequest, containing
+	// the 'NACK' from envoy on rejected configs. Only ID is set in metadata.
+	// This includes all the info that envoy (client) provides.
+	TypeURLNACK = "istio.io/nack"
+
+)
+
+// InternalGen is a Generator for XDS status updates: connect, disconnect, nacks, acks
+type InternalGen struct {
+	Server *DiscoveryServer
+
+	// TODO: track last N Nacks and connection events, with 'version' based on timestamp.
+	// On new connect, use version to send recent events since last update.
+}
+
+func (sg *InternalGen) OnConnect(node *core.Node) {
+	sg.startPush(TypeURLConnections, []*any.Any{util.MessageToAny(node)})
+}
+
+func (sg *InternalGen) OnDisconnect(node *core.Node) {
```

Comment:
nit: I think our code should rely on istio objects more than envoy where possible. Would recommend taking a model.proxy here, then we can just call proxy.node to get the core.Node from it. This way if/when we add auto registration of XDS clients as endpoints we won't need to refactor

Reply:
That's a bit controversial. 
model.Proxy is not a proto - so we can't really use it. We could define a proto - which is one of the possible long term paths. 

But at the same time this is XDS - and maybe the 'connect'/disconnect should actually be XDS top level objects instead of istio specific. There is nothing special about Istio in this event. 

That was  the main reason I used 'node' - any XDS server can implement this without taking a dependency on Istio. 

I was thinking even long term if we define the Istio specific object - we can still support the type URL based on envoy object. And ideally a top level envoy type could be defined that would be used instead of Node.

Reply:
You can still make OnDisconnect accept a model.Proxy though but send only the core.Node. The reason I suggested this is because its needed for the auto-registration. Agreed actually sending XDS/UDPA objects on the wire is probably a good idea, at least for now

Reply:
Got it, sounds good - will make the change.